app/queries/reasons_for_rejection_count_query.rb

```diff
@@ -0,0 +1,99 @@
+class ReasonsForRejectionCountQuery
+  THIS_MONTH = 'this_month'.freeze
+  BEFORE_THIS_MONTH = 'before_this_month'.freeze
+
+  Result = Struct.new(:all_time, :this_month, :sub_reasons)
+
+  def reason_counts
+    rows = ActiveRecord::Base.connection.exec_query(
+      reason_counts_sql,
+      'SQL',
+      [[nil, Time.zone.now.beginning_of_month]],
+    ).to_a
+
+    to_results(rows)
+  end
+
+  def sub_reason_counts
+    results = reason_counts
+
+    rows = ActiveRecord::Base.connection.exec_query(
+      sub_reason_counts_sql,
+      'SQL',
+      [[nil, Time.zone.now.beginning_of_month]],
+    ).to_a
+
+    add_sub_results(results, rows)
+  end
+
+private
+
+  MAPPING = {
+    candidate_behaviour_what_did_the_candidate_do: :candidate_behaviour_y_n,
+    quality_of_application_which_parts_needed_improvement: :quality_of_application_y_n,
+    qualifications_which_qualifications: :qualifications_y_n,
+    honesty_and_professionalism_concerns: :honesty_and_professionalism_y_n,
+    safeguarding_concerns: :safeguarding_y_n,
+  }.with_indifferent_access
+
+  def to_results(rows)
+    rows.reduce(ActiveSupport::HashWithIndifferentAccess.new { |hash, reason| hash[reason] = Result.new(0, 0, ActiveSupport::HashWithIndifferentAccess.new { |sub_hash, sub_reason| sub_hash[sub_reason] = Result.new(0, 0, nil) }) }) { |results, row|
+      if row['time_period'] == THIS_MONTH
+        results[row['key']].this_month += row['count'].to_i
+      end
+      results[row['key']].all_time += row['count'].to_i
+      results
+    }.with_indifferent_access
```

Comment:
A few suggestions for clarity:

Maybe something like
```
    empty_results_hash =
      ActiveSupport::HashWithIndifferentAccess.new do |hash, reason|
        hash[reason] = Result.new(
          0, 0, ActiveSupport::HashWithIndifferentAccess.new do |sub_hash, sub_reason|
            sub_hash[sub_reason] = Result.new(0, 0, nil)
          end
        )
      end
```
which can then be passed to `reduce`.

Also, perhaps `each_with_object` instead of `reduce`?

Finally, is the last call to `with_indifference_access` required?

Reply:
Much better!